### PR TITLE
feat(aws): Add AWS RAM resources

### PR DESCRIPTION
**Pull request by candiduslynx**
Add AWS RAM resources (#1984)

### CODE REVIEW

File: plugins/source/aws/codegen/recipes/ram.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+					Type:     schema.TypeString,
+					Resolver: `schema.PathResolver("ServiceName")`,
+					Options:  pk,
+				},
+			),
+		},
+	}
+	for _, r := range resources {
+		r.Service = "ram"
+		r.Description = "https://docs.aws.amazon.com/ram/latest/APIReference/API_" + r.StructName() + ".html"
```
Comment (disq) [inline plugins/source/aws/codegen/recipes/ram.go:161]: Why not get rid of `mx` above and put it as a single line here?

**Review by disq — COMMENTED** [review]
LGTM with single nit

File: plugins/source/aws/codegen/recipes/ram.go (hunk 1 above)
Comment (candiduslynx) [reply plugins/source/aws/codegen/recipes/ram.go:161]: `mx` is not used for the relations (naturally, you don't need to do multiplexing as the parent will reside in the same region).

File: plugins/source/aws/codegen/recipes/base.go [hunk 2]
```diff
@@ -277,9 +280,3 @@ func (r *Resource) generateMockTest(dir string) error {
 func (r Resource) StructName() string {
 	return reflect.TypeOf(r.Struct).Elem().Name()
 }
-
-// CloudQueryServiceName is used for accessing 'client.Services().{{.CloudqueryServiceName}}' in templates
-func (r Resource) CloudQueryServiceName() string {
-	csr := caser.New()
-	return csr.ToPascal(r.Service)
-}
```
Comment (hermanschaaf) [inline plugins/source/aws/codegen/recipes/base.go:285]: Maybe a small thing, but I think I'd rather keep this function so the logic is in one place, rather than spread across all the templates. As we saw recently, this does change from time to time. :)

**Review by hermanschaaf — CHANGES_REQUESTED** [review]
Looks good 👍  Just some minor nits / questions.

File: plugins/source/aws/docs/tables/aws_ram_principals.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+## Columns
+| Name          | Type          |
+| ------------- | ------------- |
+|_cq_source_name|String|
+|_cq_sync_time|Timestamp|
+|_cq_id|UUID|
+|_cq_parent_id|UUID|
+|account_id|String|
+|region|String|
+|id (PK)|String|
```
[the remark this answers is not in the document]
Comment (bbernays) [reply plugins/source/aws/docs/tables/aws_ram_principals.md:17]: Unless we can find documentation explicitly stating it is globally unique, we should assume that it is only unique within  region + account

File: plugins/source/aws/codegen/recipes/base.go [hunk 4]
```diff
@@ -39,6 +39,9 @@ type Resource struct {
 	// --------------------------------
 	ShouldGenerateResolverAndMockTest bool
 	ResolverAndMockTestTemplate       string
+
+	// To be used in list/paginator resolver
+	MaxResults int
```
Comment (bbernays) [inline plugins/source/aws/codegen/recipes/base.go:44]: Do we want to split the codegen changes into a separate PR?

Comment (candiduslynx) [reply plugins/source/aws/codegen/recipes/base.go:45]: I specifically added the new templates to gen the new resources, so maybe leave in single PR?
If you feel strongly it should be split I can do it, though.

**Review by hermanschaaf — APPROVED** [review]
Love the new PKColumns option 😍

File: plugins/source/aws/docs/tables/aws_ram_resource_share_invitations.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# Table: aws_ram_resource_share_invitations
+
+https://docs.aws.amazon.com/ram/latest/APIReference/API_ResourceShareInvitation.html
+
+The primary key for this table is **resource_share_invitation_arn**.
```
Comment (hermanschaaf) [inline plugins/source/aws/docs/tables/aws_ram_resource_share_invitations.md:5]: Could we rename the `resource_share_invitation_arn` column to `arn` please, to be consistent with other tables :bow:

**Review by hermanschaaf — CHANGES_REQUESTED** [review]
Sorry, just noticed two ARNs column names that should ideally be consistent with other AWS tables.

File: plugins/source/aws/docs/tables/aws_ram_resource_shares.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# Table: aws_ram_resource_shares
+
+https://docs.aws.amazon.com/ram/latest/APIReference/API_ResourceShare.html
+
+The primary key for this table is **resource_share_arn**.
```
[the remark this answers is not in the document]
Comment (candiduslynx) [reply plugins/source/aws/docs/tables/aws_ram_resource_shares.md:5]: will do

File: plugins/source/aws/codegen/recipes/base.go [hunk 7]
```diff
@@ -28,17 +28,24 @@ type Resource struct {
...
+	PKColumns             []string
 	Table                 *codegen.TableDefinition
 	Multiplex             string
 	PreResourceResolver   string
 	PostResourceResolver  string
 	Relations             []string
 	UnwrapEmbeddedStructs bool
 
+	// NameTransformer custom name transformer for resource
+	NameTransformer func(field reflect.StructField) (string, error)
```
Comment (hermanschaaf) [inline plugins/source/aws/codegen/recipes/base.go:40]: I'm aware we have the same thing in the GCP plugin, but it seems a bit like overkill to have a generic transformer function here, when (I'm pretty sure) `map[string]string` would suffice for all cases we know of in AWS 😅 The use case in GCP is also slightly different: it's for overriding things like "CORS" with "cors" instead of "c_o_r_s" (and I think we actually have better mechanisms for handling this now)

Could we maybe make it something like `OverrideColumnNames: map[string]string{}` with the key being the struct field, and the value being the new column name? So then generally it would be used something like:

```
OverrideColumnNames: map[string]string{
    "ResourceShareArn": "arn",
}
```

This also has the advantage that you don't need to guess what the transformed column name will be, you can use the struct field name, which is certain.

What do you think?

**Review by hermanschaaf — APPROVED** [review]

**Review by bbernays — APPROVED** [review]

File: plugins/source/aws/codegen/recipes/base.go (hunk 7 above)
Comment (candiduslynx) [reply plugins/source/aws/codegen/recipes/base.go:40]: If we consider the field names as keys, only the name transformer option can help us at the moment.
We can do a simple map if we use the resulting column names instead, but for now I think it works better.
